Improve latency, fix rectangular "tearing", better performance metrics with frame pacing

yshui opened the pull request:
## Latency

Rendering a frame too early can cause long latency. Generally we want to render a frame as late as possible, as long as it finishes in time before the vblank period.

## Tearing

Sometimes user will see rectangular shaped tearing when `use-damage` is enabled. This is because we started rendering early, before all the damages for that frame has arrived. Resulting in only part of an updated window being presented on to the screen.

## Metrics

Previously, we never measured how long each frame takes to render. We need this information to pace the frames properly, i.e. to achieve the things listed above. So as a nice side effect, we now know exactly how inefficient we actually are :smiling_face_with_tear: 

# Known problem

* Need to detect dpms turning screen off, which would throw off our refresh rate estimations

yshui committed:
backend: remove unused parameter 'ignore_damage'

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: collect frame timing information.

Needed to estimate refresh rate, and for us to stay in phase with
vblank.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

absolutelynothelix commented on src/backend/backend.c:
```diff
@@ -55,6 +60,59 @@ region_t get_damage(session_t *ps, bool all_damage) {
+		// damage.
+		pixman_region32_init(reg_paint);
+		return;
+	}
+
+	/// The adjusted damaged regions
+	assert(ps->o.blur_method != BLUR_METHOD_INVALID);
+	if (ps->o.blur_method != BLUR_METHOD_NONE && ps->backend_data->ops->get_blur_size) {
+		int blur_width, blur_height;
+		ps->backend_data->ops->get_blur_size(ps->backend_blur_context,
+		                                     &blur_width, &blur_height);
+
+		// The region of screen a given window influences will be smeared
+		// out by blur. With more windows on top of the given window, the
+		// influences region will be smeared out more.
+		//
+		// Also, blurring requires data slightly outside the area that needs
+		// to be blurred. The more semi-transparent windows are stacked on top
+		// of each other, the larger the area will be.
+		//
+		// Instead of accurately calculate how much bigger the damage
+		// region will be because of blur, we assume the worst case here.
+		// That is, the damaged window is at the bottom of the stack, and
+		// all other windows have semi-transparent background
+		int resize_factor = 1;
+		if (stacking_rank) {
+			resize_factor = stacking_rank;
+		}
+		// TODO(yshui): maybe we don't need to resize reg_damage, only reg_paint?
+		resize_region_in_place(reg_damage, blur_width * resize_factor,
+		                       blur_height * resize_factor);
+		*reg_paint = resize_region(reg_damage, blur_width * resize_factor,
+		                           blur_height * resize_factor);
+		pixman_region32_intersect(reg_paint, reg_paint, &ps->screen_reg);
+		pixman_region32_intersect(reg_damage, reg_damage, &ps->screen_reg);
+	} else {
+		pixman_region32_init(reg_paint);
+		pixman_region32_copy(reg_paint, reg_damage);
+	}
+}
```
am i missing something or you can replace all these branches with
```c
if (ps->o.blur_method == BLUR_METHOD_NONE || !ps->backend_data->ops->get_blur_size) {
    pixman_region32_init(reg_paint);
    if (!pixman_region32_not_empty(reg_damage)) {
        pixman_region32_copy(reg_paint, reg_damage);
    }

    return;
}
```

yshui replied:
what do you mean?

absolutelynothelix replied:
```c
void get_resized_damage(session_t *ps, int stacking_rank, bool whole_screen,
                        region_t *reg_damage, region_t *reg_paint) {
	*reg_damage = get_damage(ps, whole_screen);

    if (ps->o.blur_method == BLUR_METHOD_NONE || !ps->backend_data->ops->get_blur_size) {
        pixman_region32_init(reg_paint);
        if (pixman_region32_not_empty(reg_damage)) {
            pixman_region32_copy(reg_paint, reg_damage);
        }

        return
    }

    assert(ps->o.blur_method != BLUR_METHOD_INVALID);
    int blur_width, blur_height;
    ps->backend_data->ops->get_blur_size(ps->backend_blur_context,
                                            &blur_width, &blur_height);
    int resize_factor = 1;
    if (stacking_rank) {
        resize_factor = stacking_rank;
    }
    resize_region_in_place(reg_damage, blur_width * resize_factor,
                            blur_height * resize_factor);
    *reg_paint = resize_region(reg_damage, blur_width * resize_factor,
                                blur_height * resize_factor);
    pixman_region32_intersect(reg_paint, reg_paint, &ps->screen_reg);
    pixman_region32_intersect(reg_damage, reg_damage, &ps->screen_reg);
}
```
p.s. removed comments to keep it short

absolutelynothelix commented on src/backend/backend.c:
```diff
@@ -31,12 +31,17 @@ struct backend_operations *backend_list[NUM_BKEND] = {
+	}
 
-	if (all_damage) {
+	if (whole_screen) {
 		buffer_age = -1;
 	}
```
isn't it better to assume that `buffer_age` is `-1` and call buffer age function only if `whole_screen` is `false`?

absolutelynothelix commented on src/backend/backend.c:
```diff
@@ -55,6 +60,59 @@ region_t get_damage(session_t *ps, bool all_damage) {
+		// region will be because of blur, we assume the worst case here.
+		// That is, the damaged window is at the bottom of the stack, and
+		// all other windows have semi-transparent background
+		int resize_factor = 1;
+		if (stacking_rank) {
+			resize_factor = stacking_rank;
+		}
```
`int resize_factor = stacking_rank || 1;`?

yshui replied:
No, this is more difficult to understand.

yshui replied:
Also this is not how C's `||` operator works

absolutelynothelix replied:
oh, i see, my bad

yshui committed:
core: frame pacing

Use frame timing and render time statistic to pace frames.

Right now the criteria are simple:

* Don't render multiple frames in one vblank cycle. Otherwise the
  rendered frame will be delay multiple cycles, which isn't ideal.
* Start rendering as late as possible while still hitting vblank.

Refresh rate is estimated from a rolling average of frame timing. Render
time is predicted from the rolling maximum of past 128 frames. The
window size still needs to be investigated.

Remove glFinish calls and GL_MaxFramesAllowed=1, frame pacing superseeds
them.

Professionals might laugh at how rudimentary this is, but hopefully this
is better than what we had before. Which is absolutely nothing at all.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
Frame times don't look good. It is ~5ms/frame here. Need to investigate what's so slow.

yshui committed:
backend: add last_render_time interface.

Used for querying how long render takes to complete, used for frame
pacing.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
I wonder how well this works for NVIDIA users. I would be surprised if this works for them at all :thinking:

absolutelynothelix commented:
> I wonder how well this works for NVIDIA users. I would be surprised if this works for them at all 🤔

if there is no time limit for this pull request to be merged i can test it in maybe 5 days on an average nvidia gpu (gtx 1650). unfortunately, i’m ill now and my pc is not with me.

also, is it possible to test/debug picom remotely? i think i could give you remote access to my pc from time to time via rdp or something like this. if you’re interested you can contact me in telegram with the same login.

yshui commented:
for me, if i can get the render time to reasonable levels (`transparent-clipping` helps, because there's a lot less to draw), the latency feels pretty low, it's actually quite nice.

yshui commented:
With dual_kawase blur strength 20, the render time goes up to ~11ms :smiling_face_with_tear:

tryone144 commented:
> I wonder how well this works for NVIDIA users. I would be surprised if this works for them at all :thinking:

At least on my GTX 1070 with driver 515.43.04 it compiles and draws fine. :shrug: Just a very preliminary test though.

<details><summary>Logs</summary>

```
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_SGI_video_sync.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_SGI_swap_control.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Missing GLX extension GLX_OML_sync_control.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Missing GLX extension GLX_MESA_swap_control.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_EXT_swap_control.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_EXT_texture_from_pixmap.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_ARB_create_context.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_EXT_buffer_age.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Found GLX extension GLX_ARB_create_context_robustness.
[ 14.12.2022 22:02:43.318 glx_has_extension INFO ] Missing GLX extension GLX_MESA_query_renderer.
[ 14.12.2022 22:02:43.755 glGetUniformLocationChecked INFO ] Failed to get location of uniform 'time'. This is normal when using custom shaders.
[ 14.12.2022 22:02:43.757 gl_init INFO ] Using back buffer format 0x8051
[ 14.12.2022 22:02:43.757 gl_has_extension INFO ] Missing GL extension GL_GREMEDY_string_marker.
[ 14.12.2022 22:02:43.757 gl_init INFO ] GL vendor is NVIDIA, enable xrender sync fence.
[ 14.12.2022 22:02:43.757 gl_has_extension INFO ] Missing GL extension GL_EXT_EGL_image_storage.
[ 14.12.2022 22:02:43.757 gl_create_kernel_blur_context INFO ] Uniform locations: 3 2 1 0 4
[ 14.12.2022 22:02:43.757 gl_create_kernel_blur_context INFO ] Uniform locations: 3 2 1 0 4
```
followed by the occasional `A frame is still being rendered, not queueing redraw.` warning.

</details>

yshui commented:
@tryone144 nice! i was worried that NVIDIA's present extension implementation might use a different time base from CLOCK_MONOTONIC.

yshui committed:
core: workaround X present event quirk

When the screen turns off, X sometimes sends present complete notify for
the same frame multiple times, or even events with invalid msc/ust
number.

This will cause us to ignore it and not send a subsequent NotifyMsc
request, causing the complete notify to stop.

Now we send NotifyMsc regardless to keep the events going.

Also detect when the complete notifies skip frames, divide the interval
by frame count to estimate frame time in that case.

Upstream bug report:
https://gitlab.freedesktop.org/xorg/xserver/-/issues/1418

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: better frame pacing function

Details explained in the comments on schedule_render().

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: stop sending NotifyMsc if frame pacing is disabled

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: only check present timer alignment in the first frame

Timer can sometimes drift randomly, like when the system is suspended.
we don't want to disable frame pacing for that

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
hmph, there is a raging memory leak somewhere i need to fix.

yshui committed:
core: don't fully trust msc from present complete notify

Sometimes X sends bogus complete notifies with invalid msc number.
Instead use a saved msc number to get the next complete notify.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

absolutelynothelix commented on src/picom.c:
```diff
@@ -1382,16 +1398,58 @@ static void unredirect(session_t *ps) {
+		if (ev->event != ps->present_event_id) {
+			continue;
+		}
+		if (ev->evtype == XCB_PRESENT_EVENT_COMPLETE_NOTIFY) {
```
could this if also include if above? i mean
```c
...
if (ev->event == ps->present_event_id && ev->evtype == XCB_PRESENT_EVENT_COMPLETE_NOTIFY) {
...
```

yshui replied:
i updated the comments, hopefully it's clearer now.

absolutelynothelix replied:
i see. btw, i was looking through commits from the first to the last one and some things were already changed in process, so some of my comments may be not relatable anymore.

absolutelynothelix commented on src/picom.c:
```diff
@@ -1445,15 +1519,54 @@ static void handle_present_events(session_t *ps) {
+				drift = cne->ust - now_usec;
+			} else {
+				drift = now_usec - cne->ust;
+			}
```
good ol' `abs`? or, if you feel *mysterious* today, https://graphics.stanford.edu/~seander/bithacks.html#IntegerAbs

yshui replied:
these 2 numbers are unsigned, so can't use `abs` here :smiling_face_with_tear:

absolutelynothelix commented on src/picom.c:
```diff
@@ -1445,15 +1519,54 @@ static void handle_present_events(session_t *ps) {
+				ps->frame_pacing = false;
+				// We could have deferred a frame in queue_redraw()
+				// because of frame pacing. Unconditionally queue a frame
+				// for sim1plicity.
```
a typo

yshui replied:
thanks, fixed!

absolutelynothelix commented:
works and looks ~unclear~ good to me.

frame time doesn't look good for me too and i'm not sure why. though i have an idea that `--log-level=TRACE` constantly writes a lot of information (even in file) and may slower the rendering process?..

also, just a humble thought: is is possible to add an overlay like rtss' one? i think it's better than the wall of text in the log.
![xd](https://cms.capframex.com/api/assets/capframex-com/484f5b44-3694-4d5f-9347-232bf31871e7/grafik.png)

yshui commented:
@absolutelynothelix 

> frame time doesn't look good for me

this "frame time" is just `1s / fps`. for rendering performance, search for "Last render call took"

yshui commented:
added some initialization in `redirect_start()`

absolutelynothelix commented:
> for rendering performance, search for "Last render call took"

🥴
```
workstation:~/Downloads/picom$ grep "Last render call took" picom.log | head -n5
[ 12/21/2022 21:42:06.781 schedule_render TRACE ] Last render call took: 1319 us, rolling_max: -2147483648 us
[ 12/21/2022 21:42:06.805 schedule_render TRACE ] Last render call took: 1355 us, rolling_max: 1319 us
[ 12/21/2022 21:42:06.815 schedule_render TRACE ] Last render call took: 1314 us, rolling_max: 1355 us
[ 12/21/2022 21:42:06.832 schedule_render TRACE ] Last render call took: 1322 us, rolling_max: 1355 us
[ 12/21/2022 21:42:06.864 schedule_render TRACE ] Last render call took: 1320 us, rolling_max: 1355 us
workstation:~/Downloads/picom$ grep "Last render call took" picom.log | tail -n5
[ 12/21/2022 21:42:18.133 schedule_render TRACE ] Last render call took: 32878 us, rolling_max: 42095 us
[ 12/21/2022 21:42:18.181 schedule_render TRACE ] Last render call took: 17600 us, rolling_max: 42095 us
[ 12/21/2022 21:42:18.231 schedule_render TRACE ] Last render call took: 16849 us, rolling_max: 42095 us
[ 12/21/2022 21:42:18.281 schedule_render TRACE ] Last render call took: 15775 us, rolling_max: 42095 us
[ 12/21/2022 21:42:18.331 schedule_render TRACE ] Last render call took: 15775 us, rolling_max: 42095 us
```

absolutelynothelix commented:
if you're interested

command line: `$ build/src/picom --log-level=TRACE --log-file=picom.log`
picom.log: http://0x0.st/o5jy.log
picomrc: http://0x0.st/o5jt.txt

tryone144 commented:
@absolutelynothelix 
```console
$ tail -f picom.log | perl -nle 'BEGIN{$|=1} if (/Last render call took:/) { s/^.*took:\s(\d+\sus),.*max:\s(\d+\sus).*$//; printf "\r%s, max: %s  ", $1, $2; }'
```
for a live, single-line output in your terminal :wink:

yshui commented on src/picom.c:
```diff
@@ -207,10 +207,112 @@ static inline struct managed_win *find_win_all(session_t *ps, const xcb_window_t
+}
+
 void queue_redraw(session_t *ps) {
+	// Whether we have already rendered for the current frame.
+	// If frame pacing is not enabled, pretend this is false.
```
leftover comments no longer apply

yshui commented:
how do we implement last_render_time for xrender? maybe i can use a xsync fence for a rough estimate. :thinking:

yshui commented:
problem with xsync fences is that we can't be asynchronously notified of their completion. 

```
AwaitFence 

 When this request is executed, the processing of further requests for the client is blocked until one or more of the fences in fence-list reaches the triggered state.  If any of the fences are already in the triggered state, request processing resumes immediately. A Fence error is generated if any member of fence-list does not name a valid fence.
```

yshui commented:
i am still seeing screen freezes after DPMS off :\

absolutelynothelix commented:
@yshui, what do you mean?

yshui commented:
@absolutelynothelix so X turns your screen off after some inactivity, sometimes the screen freezes after it's turned back on

absolutelynothelix commented:
@yshui, does this happen when screen is turned off manually via `xset`?

yshui commented:
@absolutelynothelix i am not sure.

absolutelynothelix commented:
@yshui, i’ve never encountered this, but i’ll try to reproduce it later

tryone144 commented:
> i am still seeing screen freezes after DPMS off :\

I've had this happen about twice on my system now running this branch daily. Seem to be random however and I couldn't trigger it with manually forcing dpms suspend via xset.

@yshui do you have a hunch what could cause this or if a trace log could help narrow this down?

absolutelynothelix commented:
unfortunately i can’t test things like it’s my daily driver because i’m not using linux all the time. i’m mostly capable of synthetic tests trying different things and scenarios.

regarding this issue i’d try running e.g. glxgears and a script that will call xset to suspend and awake display over and over again to see if this issue can be caught this way. maybe you guys can try it right now.

yshui commented:
@tryone144 X server sends weird PresentCompleteNotify events when screen is off. This is code for this but maybe I didn't cover all cases.

Otherwise it could be some initialization issues around redirect/unredirect. Trace logging would help but it's difficult to reproduce which means there would be a ton of logs.

absolutelynothelix commented:
i was running `while true; do xset dpms force standby; sleep 5; xset dpms force on; sleep 5; done` and from time to time there was a randomly appearing glx error (i was doing nothing except staring at the screen), here it happened two times in a row. could it be the issue?
```
[ 12/24/2022 20:21:06.286 x_log_error WARN ] X error 9 DRAWABLE request 156 minor 4 serial 124938
[ 12/24/2022 20:21:06.286 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:11.533 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:11.540 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 23 serial 126153
[ 12/24/2022 20:21:16.804 x_log_error WARN ] X error 8 MATCH request 152 minor 22 serial 126266
[ 12/24/2022 20:21:16.804 x_log_error WARN ] X error 9 DRAWABLE request 156 minor 4 serial 126267
[ 12/24/2022 20:21:16.804 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:22.017 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:22.024 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 23 serial 127353
[ 12/24/2022 20:21:27.307 x_log_error WARN ] X error 8 MATCH request 152 minor 22 serial 127462
[ 12/24/2022 20:21:27.307 x_log_error WARN ] X error 9 DRAWABLE request 156 minor 4 serial 127463
[ 12/24/2022 20:21:27.307 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:27.335 glx_bind_pixmap ERROR ] GLX error at line 452: GL_INVALID_OPERATION
[ 12/24/2022 20:21:32.523 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:32.530 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 23 serial 128734
[ 12/24/2022 20:21:37.754 x_log_error WARN ] X error 8 MATCH request 152 minor 22 serial 128845
[ 12/24/2022 20:21:37.754 x_log_error WARN ] X error 9 DRAWABLE request 156 minor 4 serial 128846
[ 12/24/2022 20:21:37.754 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:37.771 glx_bind_pixmap ERROR ] GLX error at line 452: GL_INVALID_OPERATION
[ 12/24/2022 20:21:43.029 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:43.039 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 23 serial 130128
[ 12/24/2022 20:21:48.272 x_log_error WARN ] X error 8 MATCH request 152 minor 22 serial 130219
[ 12/24/2022 20:21:48.272 x_log_error WARN ] X error 9 DRAWABLE request 156 minor 4 serial 130220
[ 12/24/2022 20:21:48.272 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:53.527 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
[ 12/24/2022 20:21:53.533 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 23 serial 131400
[ 12/24/2022 20:21:58.712 x_log_error WARN ] X error 8 MATCH request 152 minor 22 serial 131493
[ 12/24/2022 20:21:58.712 x_log_error WARN ] X error 9 DRAWABLE request 156 minor 4 serial 131494
[ 12/24/2022 20:21:58.712 x_log_error WARN ] X error 161 GLX_BAD_PIXMAP request 152 minor 16 serial 0
```

yshui commented:
> i looked what kind of presentcompletenotify events

yeah there are some inconsistency across drivers. nvidia seems to be doing better in this regard. xf86-video-amdgpu is doing less well.

yshui commented:
I haven't seen the screen freeze problem for a while. I think I will merge this soonish, unless something bad is discovered.

absolutelynothelix commented on src/backend/backend.h:
```diff
@@ -292,6 +292,14 @@ struct backend_operations {
+	/// Get the render time of the last frame. If the render is still in progress,
+	/// returns false. The time is returned in `ts`. Frames are delimited by the
+	/// present() calls. i.e. after a present() call, last_render_time() should start
+	/// reporting the time of the just presen1ted frame.
```
a typo.

absolutelynothelix commented on src/backend/gl/gl_common.c:
```diff
@@ -800,6 +805,9 @@ uint64_t gl_get_shader_attributes(backend_t *backend_data attr_unused, void *sha
 }
 
 bool gl_init(struct gl_data *gd, session_t *ps) {
+	glGenQueries(2, gd->frame_timing);
```
i don't see a matching `glDeleteQueries` call. i suppose it should be in the `gl_deinit` function.

absolutelynothelix commented on src/backend/gl/gl_common.c:
```diff
@@ -945,7 +953,7 @@ bool gl_init(struct gl_data *gd, session_t *ps) {
 	log_debug("GL_VENDOR = %s", vendor);
 	if (strcmp(vendor, "NVIDIA Corporation") == 0) {
-		log_info("GL vendor is NVIDIA, don't use glFinish");
+		log_info("GL vendor is NVIDIA, enable xrender sync fence.");
```
i would add dashes in the option name (i.e. `enable xrender-sync-fence`, so it's clearer that we're talking about the corresponding option.

update: actually, i'd reword the second part as `xrender-sync-fence is enabled`, so it's clearer that we're doing something and it's not an offer to a user.

absolutelynothelix commented on src/backend/gl/gl_common.c:
```diff
@@ -1154,10 +1162,33 @@ void gl_present(backend_t *base, const region_t *region) {
+	}
+
+	GLuint64 time;
+	glGetQueryObjectui64v(gd->frame_timing[gd->current_frame_timing ^ 1],
```
a note from https://www.khronos.org/opengl/wiki/Query_Object: `Note that implementations are allowed to return 0 for any of these values. This means that the query type contains no actual information, and the implementation does not really support it.`

however, i don't think that it's possible to meet such a funny opengl implementation in the real word, so let it be just a note.

absolutelynothelix commented on src/picom.c:
```diff
@@ -207,10 +207,112 @@ static inline struct managed_win *find_win_all(session_t *ps, const xcb_window_t
+
+	ps->target_msc = ps->last_msc + frame_delay;
+	auto delay = deadline - (uint64_t)render_time_estimate - now_us;
+	delay_s = (double)delay / 1000000.0;
```
a scientific notation could be used here.

absolutelynothelix commented on src/picom.c:
```diff
@@ -1341,6 +1443,40 @@ static bool redirect_start(session_t *ps) {
 		pixman_region32_init(&ps->damage_ring[i]);
 	}
 
+	ps->frame_pacing = true;
+	if (ps->o.legacy_backends || ps->o.benchmark ||
+	    !ps->backend_data->ops->last_render_time) {
+		// Disable frame pacing if we are using a legacy backend or if we are in
+		// benchmark mode, or if the backend doesn't report render time
+		log_info("Disabling frame pacing.");
+		ps->frame_pacing = false;
+	}
+
+	if (ps->present_exists && ps->frame_pacing) {
+		ps->present_event_id = x_new_id(ps->c);
+		auto select_input = xcb_present_select_input(
+		    ps->c, ps->present_event_id, session_get_target_window(ps),
+		    XCB_PRESENT_EVENT_MASK_COMPLETE_NOTIFY);
+		auto notify_msc =
+		    xcb_present_notify_msc(ps->c, session_get_target_window(ps), 0, 0, 1, 0);
+		set_cant_fail_cookie(ps, select_input);
+		set_cant_fail_cookie(ps, notify_msc);
+		ps->present_event = xcb_register_for_special_xge(
+		    ps->c, &xcb_present_id, ps->present_event_id, NULL);
+
+		// Initialize rendering and frame timing statistics, and frame pacing
+		// states.
+		ps->last_msc_instant = 0;
+		ps->last_msc = 0;
+		ps->last_render = 0;
+		ps->target_msc = 0;
+		rolling_avg_reset(ps->frame_time);
+		rolling_max_reset(ps->render_stats);
+	} else if (ps->frame_pacing) {
+		log_error("Present extension is not supported, frame pacing disabled.");
+		ps->frame_pacing = false;
+	}
```
maybe it's better to check for the x present extension in the first `if` that disables the frame pacing? in this case, the `if-else-if` will turn into a simple `if` checking if the frame pacing left enabled after the "disabling" `if`.

absolutelynothelix commented on src/picom.c:
```diff
@@ -1382,16 +1518,107 @@ static void unredirect(session_t *ps) {
+		return;
+	}
+
+	bool event_is_invalid = false;
+	if (ps->frame_pacing) {
+		auto next_msc = cne->msc + 1;
+		if (cne->msc <= ps->last_msc || cne->ust == 0) {
+			// X sometimes sends duplicate/bogus MSC events, don't
+			// use the msc value. Also ignore these events.
+			//
+			// See:
+			// https://gitlab.freedesktop.org/xorg/xserver/-/issues/1418
+			next_msc = ps->last_msc + 1;
+			event_is_invalid = true;
+		}
+		auto cookie = xcb_present_notify_msc(ps->c, session_get_target_window(ps),
+		                                     0, next_msc, 0, 0);
+		set_cant_fail_cookie(ps, cookie);
+	}
+	if (event_is_invalid) {
+		return;
+	}
```
it could be a single `if`.

yshui replied:
you mean like move the second if inside the first one?

absolutelynothelix replied:
yeah, both variable definition and check if it's true could be in the `if (ps->frame_pacing)`.

absolutelynothelix commented:
can't check if it builds and works for me now, but i've looked through the code and it looks good to me. i left some suggestions which you may apply if you find them appropriate.

absolutelynothelix commented:
oh, and one more thing that i forgot to check when i was reviewing: looks like these changes eliminate the need in the `is_nvidia` variable, but it wasn't removed.

yshui commented:
Hmm, I've been seeing explosive memory usage growth recently, seemingly only happens while screen is off.

Results from valgrind/LSan don't immediately make sense, I am still looking into it.

absolutelynothelix commented:
@yshui, there is a resource leak that i’ve mentioned above. in theory (can’t test right now), it may show itself when unredirecting and redirecting back rapidly (basically hold f11 in an application that makes unredirection possible if in fullscreen), but it shouldn’t cause “explosive memory usage growth”. take a look at the `xrestop` output, maybe?

yshui commented:
@absolutelynothelix I think I may have actually found a bug in mesa.

yshui commented:
Yep it's a bug introduced in mesa 23.0-rc4

yshui committed:
core: make sure unredirection happens when screen is off

So when the screen is off, we calls queue_redraw, hoping draw_callback
will be called and unredirects the screen. However, queue_redraw doesn't
queue another redraw when one is already queued. Redraws can be queued
in two ways: one is timer based, which is fine, because it will be
triggered no matter what; the other is frame based, which is triggered
by Present events.

When the screen is off, X server, depends on the driver, could send
abnormal Present events, which we ignore. But that also means queued
frame based redraw will stop being triggered.

Those two factors combined means sometimes unredirection does not happen
when the screen is off. Which means we aren't going to free the GL
context, which are still receiving Present events, but can't handle
them, because we are not rendering anything with GL. In the end all
these causes memory usage to balloon, until the screen is turned on and
we start rendering again.

And all these is not caught by leak checkers because this technically is
not a leak, as everything is eventually freed.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
I finally caught the bug that has been been bugging me this whole time. I think I have fixed the screen freeze problem (also a memory usage growth problem). Now I am really going to merge this.

absolutelynothelix commented:
i just wanted to test if a missing `glDeleteQueries` call can really cause a noticable memory leak when (un)redirecting rapidly but found to serious issues instead :smiling_face_with_tear:

first, picom-pacing can easily freeze the entire screen and probably hang itself because i can't stop it with ctrl-c and have to go to a tty to kill it with `killall`. it's really easy to trigger: find a window that makes unredirection possible when in fullscreen and toggle it fullscreen a few times (hold f11, but in my case even clicking it is enough). the last words of picom are:
```
[ 03/08/2023 15:50:56.589 schedule_render WARN ] Delay too long: 18446744032084.671875 s, render_time: 9747 us, frame_time: 0 us, now_us: 41624869567 us, next_msc: 0 us
```

i'm not an expert, but `18446744032084.671875 s` doesn't seem right. i don't feel like 5 billion hours passed.

2) it worsens latency. in the same conditions picom-vanilla performs better than picom-pacing. there is a slightly noticable (but sill visible to an eye) latency when i move a window around.

absolutelynothelix commented:
the observations above were made on nvidia driver. i'm on nouveau now and the screen doesn't freeze anymore when (un)redirecting rapidly, but performance degradation is here as well.

*man i hate nvidia so much they didn't implement dri3 in their proprietary driver and nouveau didn't implement vulkan at all*

yshui commented:
@absolutelynothelix 

> frame_time: 0 us

Looks like `scheduler_render` is called when we don't have a frame time estimate. Can you try turning on the trace log? There should be more information there.

> second, it worsens latency

Interesting. On amdgpu here the latency is much better with pacing.

absolutelynothelix commented:
> Interesting. On amdgpu here the latency is much better with pacing.

i tried different ways of recording but the best one so far is the iphone 60fps recording, though i'm not sure that you'll be able to notice the difference. give it a try: https://sinair.ru/f/VijFFO6I?dl=

it feels like picom-vanilla always runs at 60fps while picom-pacing sometimes drops fps.

yshui commented:
> rolling_max: -2147483648 us

yeah, so i think my guess was right. i need to investigate how this could've slipped through, because we do check for this and skip pacing for that.

> it feels like picom-vanilla always runs at 60fps while picom-pacing has sudden fps drops.

this is a problem i've noticed too. the rendering time estimation is a bit too aggressive. a single frame taking too long to render would cut the frame rate in half for 1~2 seconds, because we use a rolling maximum. i need to tune the heuristics.

yshui committed:
core: check we have both frame time and render time before pacing

We only checked render time. If we don't have frame time estimates, we
would divide by zero and end up with wild scheduling delays.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
@absolutelynothelix hi, can you add a 

```c
if (frame_time == 0) {
    abort();
}
```

after `auto frame_time = ...` in `schedule_render`? and give me the stack trace when that abort happens?

i think sometimes we can have the render time estimates but not the frame time, and I need to figure out what are the scenarios.

absolutelynothelix commented:
@yshui,
```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
44            return INTERNAL_SYSCALL_ERROR_P (ret) ? INTERNAL_SYSCALL_ERRNO (ret) : 0;                                                                                                                                                        
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007fd27de17953 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007fd27ddc8ea8 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fd27ddb253d in __GI_abort () at abort.c:79
#4  0x0000563105d05eef in schedule_render (ps=0x563106634540) at ../src/picom.c:244
#5  0x0000563105d06284 in queue_redraw (ps=0x563106634540) at ../src/picom.c:288
#6  0x0000563105d371cd in ev_handle (ps=0x563106634540, ev=0x5631067ffee0) at ../src/event.c:709
#7  0x0000563105d0b73f in x_event_callback (loop=0x7fd27e3b6060, w=0x563106634540, revents=1) at ../src/picom.c:1826
#8  0x00007fd27e3ab0cb in ev_invoke_pending () from /usr/lib/libev.so.4
#9  0x00007fd27e3aed10 in ev_run () from /usr/lib/libev.so.4
#10 0x0000563105d109fd in session_run (ps=0x563106634540) at ../src/picom.c:2750
#11 0x0000563105d10d58 in main (argc=1, argv=0x7ffcc5795408) at ../src/picom.c:2856
```

my scenario is to open gimp and hold f11 with the following configuration file, happens everytime:
```
# General
backend                 = "egl";

vsync                   = true;
dithered-present        = true;
unredir-if-possible     = true;

dbus                    = true;
log-level               = "INFO";
log-file                = "/home/helix/.local/state/picom.log";

# Opacity
opacity-rule            = [
	"80:class_i = 'code-oss'"
];

# Shadow
shadow                  = true;
shadow-radius           = 8;
shadow-offset-x         = -8;
shadow-offset-y         = -8;
shadow-opacity          = 0.3;
shadow-exclude          = [
	"window_type = 'combo'",
	"window_type = 'popup_menu'",
	"window_type = 'tooltip' && class_g != 'firefox'",
	"window_type = 'utility'"
];

# Fading
fading                  = true;
fade-in-step            = 0.05;
fade-out-step           = 0.05;
fade-delta              = 3;

# Blur
blur: {
	method   = "dual_kawase";
	strength = 10;
};

blur-background-exclude = [
	"class_i = 'slop'",
	"window_type != 'normal'"
];

# Rounded corners
corner-radius           = 8;
rounded-corners-exclude = [
	"window_type != 'normal' && window_type != 'dialog'"
];
```

yshui commented:
@absolutelynothelix thanks for your help, because I couldn't reproduce this on my side, with the same setup.

i think the cause of this problem is that, the redraw is triggered by an event, which came in after the render has completed (so we have a render time), but before we received the PresentNotify. and the fix I think is to just check for `frame_time == 0` in the if. I will push that, please give it a go.

absolutelynothelix commented:
@yshui, now the freeze when unredirecting and redirecting back rapidly is gone. however, now it feels that `picom-pacing` is always noticable laggy compared to vanilla picom: https://sinair.ru/f/YsqGbA0w?dl=

yshui commented:
@absolutelynothelix thanks for confirming. now if you put a `if (frame_delay > 1) log(...)` in `schedule_render`, it would log when picom needs to skip frames. watching your video i am not entirely sure if that's what is happening... so it would help to actually confirm it.

absolutelynothelix commented:
@yshui, i put it here:
![image](https://user-images.githubusercontent.com/52252283/235261487-9e1a1ab6-c90a-4b98-ac0f-d2bd483adef8.png)

the result is https://sinair.ru/f/QM7mPKlF?dl=

yshui commented:
@absolutelynothelix seems like with blur enabled the render takes too long to fit in one frame. can you print out render_time_estimate when frame_delay > 1?

absolutelynothelix commented:
@yshui,
```diff
+       if (frame_delay > 1) {
+               log_fatal("frame_delay %lu, render_time_estimate %d", frame_delay, render_time_estimate);
+       }
```

single window, start picom, move window around:
```
...
[ 05/01/2023 15:21:28.648 gl_create_kernel_blur_context INFO ] Uniform locations: 3 2 1 0 4
[ 05/01/2023 15:21:34.359 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 17131
[ 05/01/2023 15:21:34.409 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.459 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.492 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.526 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.576 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.609 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.642 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.676 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.709 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.742 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.776 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
[ 05/01/2023 15:21:34.809 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 22105
...
```

two windows, start picom, move the second window around:
```
...
[ 05/01/2023 15:21:42.260 gl_create_kernel_blur_context INFO ] Uniform locations: 3 2 1 0 4
[ 05/01/2023 15:21:45.000 schedule_render FATAL ERROR ] frame_delay 18, render_time_estimate 3280
[ 05/01/2023 15:21:47.107 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 9362
[ 05/01/2023 15:21:47.575 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 20446
[ 05/01/2023 15:21:47.624 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 24296
[ 05/01/2023 15:21:47.675 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 27008
[ 05/01/2023 15:21:47.708 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 27008
[ 05/01/2023 15:21:47.741 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 27008
[ 05/01/2023 15:21:47.774 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 27008
[ 05/01/2023 15:21:47.808 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 27008
[ 05/01/2023 15:21:47.858 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:47.891 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:47.925 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:47.958 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:47.991 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:48.024 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:48.058 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:48.091 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
[ 05/01/2023 15:21:48.124 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 28120
...
```

yshui commented:
hmm, is dual kawase supposed to be this slow? cc @tryone144

absolutelynothelix commented:
moving two windows around with blur disabled takes me back to the previous experience:
> it feels like picom-vanilla always runs at 60fps while picom-pacing has sudden fps drops.

```
[ 05/01/2023 21:24:36.003 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 2598
[ 05/01/2023 21:24:40.452 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8815
[ 05/01/2023 21:24:40.481 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8815
[ 05/01/2023 21:24:40.718 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8815
[ 05/01/2023 21:24:41.830 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 9887
[ 05/01/2023 21:24:42.044 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 9887
[ 05/01/2023 21:24:44.687 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8383
[ 05/01/2023 21:24:44.717 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8383
[ 05/01/2023 21:24:44.749 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8383
[ 05/01/2023 21:24:46.259 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 12210
[ 05/01/2023 21:24:48.818 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 5891
[ 05/01/2023 21:24:48.966 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 5891
[ 05/01/2023 21:24:50.597 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 13883
[ 05/01/2023 21:24:50.625 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 13883
[ 05/01/2023 21:24:50.657 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 13883
[ 05/01/2023 21:24:50.891 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 13883
[ 05/01/2023 21:24:53.243 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 14445
[ 05/01/2023 21:24:53.273 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 14445
[ 05/01/2023 21:24:53.345 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 14445
[ 05/01/2023 21:24:58.478 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8674
[ 05/01/2023 21:24:58.595 schedule_render FATAL ERROR ] frame_delay 2, render_time_estimate 8674
```

yshui commented:
@absolutelynothelix that i can probably fix. but we still have to figure out if we can make blurring faster.

absolutelynothelix commented:
@yshui, i often use gl-powered backends as references for vulkan backend (that’s a bad idea, actually) and after reading multiple times some of their parts i feel like there is a plenty of room for various optimizations and simplifications, it seems kinda overengineered in some places for me.

yshui commented:
@absolutelynothelix ok. let me improve the heuristics a bit and we can merge this, and worry about improving the performance later.

absolutelynothelix commented:
iirc this patch includes that pacing is disabled if something goes wrong (can’t look at the code rn). can we make it optional but enabled by default? just like `--no-use-damage` or it means disabling vsync?

absolutelynothelix commented:
> @yshui, i often use gl-powered backends as references for vulkan backend (that’s a bad idea, actually) and after reading multiple times some of their parts i feel like there is a plenty of room for various optimizations and simplifications, it seems kinda overengineered in some places for me.

i think that happened because all the features and bug fixes were implemented on top of the existing code without rethinking it and trying to integrate new code gracefully. it allowed to introduce new features and fix bugs quick but it led to whatever we have now.
> iirc this patch includes that pacing is disabled if something goes wrong (can’t look at the code rn). can we make it optional but enabled by default? just like `--no-use-damage` or it means disabling vsync?

that's a kind of paradox. in order to make performance improvements we need some kind of metrics and this patch introduces them but also introduces some kind of performance degradation.

i think it's a good idea to get together someday with you, @tryone144, probably other contributors who are interested in working on picom to this day and discuss different approaches. irc, discord, telegram, whatever. at least i have something to say as i managed to reimplement parts of common gl code in a different, simpler and, hopefully, more performant way.

yshui commented:
@absolutelynothelix actually i just happen to be thinking of creating a discord server as well. but I have no experience running a discord server, e.g. moderation, advertising the server, etc. we could create one and see how it goes, maybe.

absolutelynothelix commented:
@yshui, actually, i’ve messaged you about a discord server about three years ago :D

i don’t think that we need to «advertise» the server, it’s a picom’s development/community server that’s about to start from discussing development strategies so putting a link/badge into README should be enough.

moderation… well, it depends how you look at things. there is the discord tos for severe violations and i don’t care about minor ones. i’m okay with what is usually prohibited. things are simple if you simply don’t care about them /shrug. however, it should be enough to create a basic set of rules and put you, collaborators and some contributors on moderation. maybe someone will even decide to contribute in such way.

EpsilonKu commented:
I can create discord server. But does moderation make sense? If u want server not only for picom (shitposting etc.). I don't think 10 ppl will send any suspect memes.

EpsilonKu commented:
Emm... so there is no comment. I decided to create server. https://discord.gg/SY5JJzPgME
Should I PR it to README? If u don't like it, u can create another.

yshui commented:
BTW, I plotted some render time:

![Figure_1](https://github.com/yshui/picom/assets/366851/7f0432f8-a8f8-4b72-b118-419f23617097)

Looks multimodal to me. A running maximum would force us to be too pessimistic (those tiny bars at the right end). But I don't know what the best strategy is. Current thought is maybe use a 99 percentile.

yshui committed:
core: factor out code for estimating the rendering time budget

Add a render_statistics type to encapsulate all the statistics done on
rendering times. And use that to estimate the time budget for rendering
and frame pacing.

Tweak the rolling window utilities a bit so we can reuse one rolling
window for multiple statistics.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: don't update render statistics if we didn't actually render

Sometimes a scheduled render can end up doing nothing, e.g. if the
damage region is empty. In that case we don't have valid data to
collect and thus shouldn't update the statistics.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: use SCHED_RR scheduling

Make picom realtime to reduce latency, and make rendering times more
predictable to help pacing.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
options: add a no-frame-pacing option

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
@absolutelynothelix hey, i made some effort to make things smoother. can you try to see if it's better for you?

absolutelynothelix commented:
@yshui, i'd say it feels pretty much same. i can record a new video if you wish.

yshui commented:
@absolutelynothelix the logs would be more useful.

absolutelynothelix commented:
> the logs would be more useful.

are you interested in particular messages or in all messages introduced with this pull request?

yshui commented:
@absolutelynothelix just throw the whole trace logs here, i know how to use grep ;)

absolutelynothelix commented:
i was moving a small terminal window from where i launched picom in front of a big terminal window with nonsense activity. i stopped picom some time after window movement started to lag, so all the interesting information should be somewhere in the end.

[picom.log](https://github.com/yshui/picom/files/11712816/picom.log)

p.s.:
* configuration file: http://0x0.st/Hc2_.txt
* command line overrides: `--log-level=TRACE --log-file=picom.log`

yshui commented:
@absolutelynothelix yeah the frames just took too long to render, that's a different problem we need to resolve aside from this PR

```
[ 06/10/2023 19:19:45.794 schedule_render TRACE ] Delay: 0.011906 s, last_msc: 23556707859, render_budget: 16658, frame_time: 16654, now_us: 23556712603, next_msc: 23556741167, target_msc: 1413461, divisor: 2
```

yshui committed:
core: print some timing info from draw_callback_impl

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
@EpsilonKu I joined! My discord username is yshui#7811, can you make me a moderator? I will also add a link to it in README.

yshui commented:
@tryone144 do you have a discord account by any chance?
